repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -238,54 +239,50 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
+        token.roles.groupBy(_.tenantId).map({ case (key, value) => key.getOrElse(DomainRoleTenantKey) -> value.map(_.name).toSet })
+    }
+
     def buildTenantHeader(defaultTenant: Option[String],
                           roleTenants: Seq[String],
-                          matchedUriTenant: Option[String]): Vector[String] = {
+                          matchedTenants: Set[String]): Vector[String] = {
       val sendAllTenants = configuration.getTenantHandling.isSendAllTenantIds
       val sendTenantIdQuality = Option(configuration.getTenantHandling.getSendTenantIdQuality)
       val sendQuality = sendTenantIdQuality.isDefined
       val defaultTenantQuality = sendTenantIdQuality.map(_.getDefaultTenantQuality).getOrElse(0.0)
-      val uriTenantQuality = sendTenantIdQuality.map(_.getUriTenantQuality).getOrElse(0.0)
+      val matchedTenantQuality = sendTenantIdQuality.map(_.getUriTenantQuality).getOrElse(0.0)
       val rolesTenantQuality = sendTenantIdQuality.map(_.getRolesTenantQuality).getOrElse(0.0)
 
       case class PreferredTenant(id: String, quality: Double)
 
-      val preferredTenant = (defaultTenant, matchedUriTenant) match {
-        case (Some(default), Some(uri)) =>
-          val quality = if (default.equals(uri)) {
-            math.max(defaultTenantQuality, uriTenantQuality)
-          } else {
-            uriTenantQuality
-          }
-          Some(PreferredTenant(uri, quality))
-        case (None, Some(uri)) =>
-          Some(PreferredTenant(uri, uriTenantQuality))
-        case (Some(default), None) =>
-          Some(PreferredTenant(default, defaultTenantQuality))
-        case (None, None) =>
-          if (roleTenants.nonEmpty) Some(PreferredTenant(roleTenants.head, rolesTenantQuality))
-          else None
-      }
+      val preferredTenants =
+        if (matchedTenants.nonEmpty) {
+          matchedTenants.map(PreferredTenant(_, matchedTenantQuality))
+        } else if (defaultTenant.nonEmpty) {
+          defaultTenant.toSet.map((tenant: String) => PreferredTenant(tenant, defaultTenantQuality))
+        } else if (roleTenants.nonEmpty) {
+          val tenantsFromRoles = roleTenants.toSet.map((tenant: String) => PreferredTenant(tenant, rolesTenantQuality))
+          if (sendAllTenants) tenantsFromRoles else tenantsFromRoles.take(1)
+        } else {
+          Set.empty
+        }
 
       if (sendAllTenants && sendQuality) {
-        val priorityTenants = (defaultTenant, matchedUriTenant) match {
-          case (Some(default), Some(uri)) => Vector(s"$default;q=$defaultTenantQuality", s"$uri;q=$uriTenantQuality")
-          case (Some(default), None) => Vector(s"$default;q=$defaultTenantQuality")
-          case (None, Some(uri)) => Vector(s"$uri;q=$uriTenantQuality")
-          case (None, None) => Vector.empty[String]
-        }
-        priorityTenants ++ roleTenants.map(tid => s"$tid;q=$rolesTenantQuality")
+        val matched = matchedTenants.map(matched => s"$matched;q=$matchedTenantQuality")
```

Comment:
Ohh, this is gross. Why are we making the header values by hand?

Reply:
The answer is that it's what we have been doing since the old auth filters.

However, I'm totally alright with changing it if you would like. I mean, I would like to if there are no objections.